source/module_esolver/esolver_sdft_pw_tool.cpp

```diff
@@ -0,0 +1,539 @@
+        const int nstogroup = GlobalV::NSTOGROUP;
+        int nrecv[nstogroup];
+        int displs[nstogroup];
+        MPI_Allgather(&totbands_per,1,MPI_INT,nrecv,1,MPI_INT,PARAPW_WORLD);
+        displs[0] = 0;
+        for(int i = 1; i < nstogroup ; ++i)
+        {
+            displs[i] = displs[i-1] + nrecv[i-1];
+        }
+        for(int i = 0 ; i < nstogroup ; ++i)
+        {
+            nrecv[i] *= npwx;
+            displs[i] *= npwx;
+        }
+#endif
+        
+
+        //-----------------------------------------------------------
+        //               sto conductivity
+        //-----------------------------------------------------------
+        //before loop
+
+        //|psi>
+        psi::Psi<std::complex<double>> psi0(1,totbands_per,npwx,GlobalC::kv.ngk.data()); //|psi>
+        psi::Psi<std::complex<double>> sfpsi0(1,totbands_per,npwx,GlobalC::kv.ngk.data()); //sqrt(f)|psi>
+        psi::Psi<std::complex<double>> hpsi0(1,totbands_per,npwx,GlobalC::kv.ngk.data()); //h|psi>
+        psi::Psi<std::complex<double>> hsfpsi0(1,totbands_per,npwx,GlobalC::kv.ngk.data()); //h*sqrt(f)|psi>
+        //j|psi> j1=p  j2=(Hp+pH)/2 - mu*p
+        psi::Psi<std::complex<double>> j1psi(1,ndim*totbands_per,npwx,GlobalC::kv.ngk.data());
+        psi::Psi<std::complex<double>> j2psi(1,ndim*totbands_per,npwx,GlobalC::kv.ngk.data());
+        //(1-f)*j*sqrt(f)|psi>
+        psi::Psi<std::complex<double>> j1sfpsi(1,ndim*totbands_per,npwx,GlobalC::kv.ngk.data());
+        psi::Psi<std::complex<double>> j2sfpsi(1,ndim*totbands_per,npwx,GlobalC::kv.ngk.data());
+        double* en;
+        if(ksbandper > 0)   en = new double [ksbandper];
+        for(int ib = 0 ; ib < ksbandper ; ++ib)
+        {
+            en[ib] = this->pelec->ekb(ik, ib+startband);
+        }
+        for(int ib = 0 ; ib < totbands_per ; ++ib )
+        {
+            std::complex<double> *tmp, *tmp2;
+            double fac = 1;
+            
+            if(ib < ksbandper)
+            {
+                tmp = &(this->psi[0](ik,ib+startband,0));
+                tmp2 = &(this->psi[0](ik,ib+startband,0));
+                fac = sqrt(stoiter.stofunc.fd(en[ib]));
+            }
+            else
+            {
+                if(GlobalV::NBANDS > 0)
+                    tmp = &stowf.chiortho[ik](ib-ksbandper,0);
+                else
+                    tmp = &stowf.chi0[ik](ib-ksbandper,0);
+                tmp2 = &stowf.shchi[ik](ib-ksbandper,0);
+            }
+        
+            for(int ig = 0; ig < npw ; ++ig)
+            {
+                psi0(ib,ig) = tmp[ig];
+                sfpsi0(ib,ig) = fac * tmp2[ig];
+            }
+        }
+        
+
+        //init j1psi,j2psi,j1sfpsi,j2sfpsi
+        for(int id = 0 ; id < ndim ; ++id)
+        {
+            for(int ib = 0 ; ib < totbands_per ; ++ib )
+            {
+                for(int ig = 0; ig < npw ; ++ig)
+                {
+                    std::complex<double> psiig = psi0(ib,ig);
```

Comment:
The use of "psi(ib, ig)"  in multi-for-loops would cause low efficiency because of the embedded action "ib*nbasis+ig", you can use lapack interface or just use pointer to avoid the "ib*nbasis+ig" action.

Reply:
ibnbasis+ig almost does not cost time.  Since we have defined psi(ib,ig), we should use it. In previous codes,  memory of psi0(ib,ig) is accessed twice, so I define "psiig = psi0(ib,ig)". But now maybe we do not need it.